File: src/lib/latency.ts
import { LatencyStatus } from "@/types";
import { LATENCY_THRESHOLDS } from "./constants";

/**
 * Classify latency into status categories
 */
export function getLatencyStatus(latencyMs: number): LatencyStatus {
  if (latencyMs <= LATENCY_THRESHOLDS.GOOD) return "good";
  if (latencyMs <= LATENCY_THRESHOLDS.FAIR) return "fair";
  return "poor";
}

/**
 * Get the color for a latency status
 */
export function getLatencyColor(status: LatencyStatus): string {
  const colors = {
    good: "#22c55e", // ping-good (green)
    fair: "#eab308", // ping-fair (yellow)
    poor: "#ef4444", // ping-poor (red)
  };
  return colors[status];
}

/**
 * Get the Tailwind class for a latency status
 */
export function getLatencyColorClass(status: LatencyStatus): string {
  const classes = {
    good: "bg-ping-good",
    fair: "bg-ping-fair",
    poor: "bg-ping-poor",
  };
  return classes[status];
}

/**
 * Get a human-readable label for a latency status
 */
export function getLatencyLabel(status: LatencyStatus): string {
  const labels = {
    good: "Excellent",
    fair: "Average",
    poor: "Poor",
  };
  return labels[status];
}

/**
 * Get the badge variant for a latency status
 */
export function getLatencyBadgeVariant(
  status: LatencyStatus
): "default" | "secondary" | "destructive" {
  const variants = {
    good: "default" as const,
    fair: "secondary" as const,
    poor: "destructive" as const,
  };
  return variants[status];
}

/**
 * Calculate the average latency from an array of pings
 */
export function calculateAverageLatency(latencies: number[]): number {
  if (latencies.length === 0) return 0;
  const sum = latencies.reduce((acc, val) => acc + val, 0);
  return Math.round(sum / latencies.length);
}

/**
 * Calculate the median latency from an array of values
 */
export function calculateMedianLatency(latencies: number[]): number {
  if (latencies.length === 0) return 0;
  const sorted = [...latencies].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 !== 0
    ? sorted[mid]
    : Math.round((sorted[mid - 1] + sorted[mid]) / 2);
}

/**
 * Calculate standard deviation (jitter) from latency samples
 */
export function calculateJitter(latencies: number[]): number {
  if (latencies.length < 2) return 0;
  const mean = latencies.reduce((a, b) => a + b, 0) / latencies.length;
  const squaredDiffs = latencies.map((val) => Math.pow(val - mean, 2));
  const variance = squaredDiffs.reduce((a, b) => a + b, 0) / latencies.length;
  return Math.round(Math.sqrt(variance));
}

File: apps/mobile/lib/domain/latency.dart
import 'dart:math' as math;

import 'constants.dart';

enum LatencyStatus { good, fair, poor }

LatencyStatus getLatencyStatus(int latencyMs) {
  if (latencyMs <= LatencyThresholds.good) return LatencyStatus.good;
  if (latencyMs <= LatencyThresholds.fair) return LatencyStatus.fair;
  return LatencyStatus.poor;
}

String getLatencyLabel(LatencyStatus status) {
  switch (status) {
    case LatencyStatus.good:
      return 'Excellent';
    case LatencyStatus.fair:
      return 'Average';
    case LatencyStatus.poor:
      return 'Poor';
  }
}

int calculateAverageLatency(List<int> latencies) {
  if (latencies.isEmpty) return 0;
  final sum = latencies.reduce((acc, val) => acc + val);
  return (sum / latencies.length).round();
}

int calculateMedianLatency(List<int> latencies) {
  if (latencies.isEmpty) return 0;
  final sorted = [...latencies]..sort();
  final mid = sorted.length ~/ 2;
  return sorted.length % 2 != 0
      ? sorted[mid]
      : ((sorted[mid - 1] + sorted[mid]) / 2).round();
}

int calculateJitter(List<int> latencies) {
  if (latencies.length < 2) return 0;
  final mean = latencies.reduce((a, b) => a + b) / latencies.length;
  final squaredDiffs = latencies.map((val) => math.pow(val - mean, 2));
  final variance = squaredDiffs.reduce((a, b) => a + b) / latencies.length;
  return math.sqrt(variance).round();
}
